src/engine.ml

```diff
@@ -1883,7 +1883,7 @@ let resolve_connect_client config ts s ev =
     | `Times m -> m > r
     | `Unlimited -> false
   in
-  let next_or_fail idx retry =
+  let next_or_fail t idx retry =
```

Comment:
Here we shadow `t`. The reason is because we return `next_or_fail`, and when it is called the state may have changed, at least judging from the structure of the code.

Reply:
Yes. It's not immediately obvious how to refactor `next_or_fail`. I arguably made things worse by passing `client` (now `make_client`) outside of the function(s) it's defined in.